### ISSUE TITLE
MKL shared libraries and Libdl on Mac/Linux

### ISSUE THREAD

**Issue by Crown421** [JuliaComputing/MKL.jl#19]
I have been updating the VML.jl package for 1.x, which makes the vectorized function in the VML library available in Julia. 
Conveniently, when one adds MKL.jl all necessary shared libraries are also added to the system, and on Windows these are also added to a path that Libdl checks, so `Libdl.dlopen(:mkl_rt)` works after installing MKL. 

However, on Mac and Linux this is apparently not the case. There, we find
```
julia> Libdl.dlopen(:libmkl_rt)
ERROR: could not load library "libmkl_core"
dlopen(libmkl_rt.dylib, 1): image not found
```
and on Linux `libmkl_rt.dylib: cannot open shared object file: No such file or directory`, after adding MKL.jl.

To fix it one can take `<path to MKL>/deps/usr/lib` and either add to (system) environmental variable `DYLD_FALLBACK_LIBRARY_PATH`, or `push!` to `Libdl.DL_LOAD_PATH` in julia, or the same for `LD_LIBRARY_PATH` on Linux.

However, I am not sure what is being done differently on Windows. I also know that it worked 'out of the box' on Linux two months ago, because at that point I built docker images where I just added MKL.jl and VML.jl and was ready to go without any adjustments in the environmental variables

Would it be reasonable on Unix based systems to make it possible (again) for Libdl to find the installed binaries? This would make using VML and thus the full range of what MKL offers very seamless.

**Comment by andreasnoack** [comment]
Couldn't you just depend directly on this package and the use our `libmkl` variable?

**Comment by aminya** [comment]
Wasn't this solved here: https://github.com/JuliaComputing/MKL.jl/issues/18

> On Windows the name of dll is :mkl_rt instead of :libmkl_rt, and so the following works:
> ```julia
> using Libdl
> dlopen(:mkl_rt)
> ```

**Comment by Crown421** [comment]
I could explicitly depend on MKL.jl, but if possible I would prefer to not and allow users to use standalone MKL/VML from intel. 
MKL.jl alters the system image, and loads a specific version of MKL, so if someone wants to use some specific version from Intel, I would rather not force them to still install MKL.jl as a dependency. 

I found a workable solution  that fairly robustly finds the path of MKL.jl if it is installed, but the functionality I used has been reworked a few time as far as I can tell, and may vanish in future Julia releases. 
Basically, it becomes difficult once either the location of the downloaded libraries within MKL.jl changes, or the ability to get the path a package without loading it first goes away. 

I can see reasons for deliberately not adding the libraries to the search path permanently, but it used to be added and it makes using the shared libraries using standard tools much easier.

**Comment by KristofferC** [comment]
Depend on MKL_jll and load it only if `MKLROOT` is not set?

**Comment by KristofferC** [comment]
> However, on Mac and Linux this is apparently not the case. 

Use this way: https://github.com/JuliaSparse/Pardiso.jl/blob/81c8391e86cebaf71e936363664d878b22b1fd9a/src/Pardiso.jl#L37-L43.

There is nothing you need to do with this package here. Just depend on MLK_jll and do something like: https://github.com/JuliaSparse/Pardiso.jl/blob/81c8391e86cebaf71e936363664d878b22b1fd9a/src/Pardiso.jl#L25-L27 together with https://github.com/JuliaSparse/Pardiso.jl/blob/81c8391e86cebaf71e936363664d878b22b1fd9a/deps/build.jl#L67-L83.
